src/cargo/util/config/mod.rs

```diff
@@ -532,6 +532,21 @@ impl Config {
                 return true;
             }
         }
+
+        self.env
+            .iter()
```

Comment:
This shouldn't iterate through the entire environment inside of every call to `has_key`; I don't think that's the right implementation. This check should occur once in the life of Cargo.

One idea would be to do this in `Config::new`, and just look for *any* environment variable that `.starts_with("CARGO_TARGET_")` and contains any lowercase letters (or contains any dashes for that matter). This would also have the advantage of catching this for all targets, not just the current target.

Reply:
I think this code only runs when all the previous checks return false. If the envs are set correctly, this code wouldn't run at all since has_key would return early. 

Moreover, when I  implemented this code in Config::new;  the warning showed up in every cargo command( check, release, test etc...) and in every project even if the warning wasn't relevant to the project. 

Also, with this implementation, the user only need to add a new correctly formatted env variable for the warning to go away. When I implemented the warning in Config::new, the user must first unset the incorrect env, then add a new one. 

Would you reconsider the implementation in has_key?